// line-readable-stream/line-readable-stream.js
///<reference path="typings/node/node.d.ts"/>
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var Stream = require("stream");

var LineReadableStream = (function (_super) {
    __extends(LineReadableStream, _super);
    /**
    * @constructor
    */
    function LineReadableStream(underlyingStream, delimiter) {
        if (typeof delimiter === "undefined") { delimiter = "\n"; }
        _super.call(this);
        if (!underlyingStream)
            throw new Error("LineReadableStream requires an underlying stream");

        this._underlyingStream = underlyingStream;
        this._data = "";
        this._delimiter = delimiter;

        this.initializeEvents();
    }
    Object.defineProperty(LineReadableStream.prototype, "underlyingStream", {
        get: function () {
            return this._underlyingStream;
        },
        enumerable: true,
        configurable: true
    });

    Object.defineProperty(LineReadableStream.prototype, "readable", {
        get: function () {
            return this._underlyingStream.readable;
        },
        enumerable: true,
        configurable: true
    });

    Object.defineProperty(LineReadableStream.prototype, "delimiter", {
        set: function (value) {
            this._delimiter = value;
        },
        enumerable: true,
        configurable: true
    });

    LineReadableStream.prototype.initializeEvents = function () {
        var _this = this;
        this._underlyingStream.on("data", function (chunk) {
            _this._data += chunk;
            var lines = _this._data.split(_this._delimiter);
            _this._data = lines.pop();
            lines.forEach(function (line) {
                return _this.emit("line", line);
            });
        });
        this.underlyingStream.on("end", function () {
            if (_this._data.length > 0) {
                var lines = _this._data.split(_this._delimiter);
                lines.forEach(function (line) {
                    return _this.emit("line", line);
                });
            }
            _this.emit("end");
        });
    };

    /**
    * Start overriding EventEmitter methods so we can pass through to underlyingStream If we get a request for an event we don't know about, pass it to the underlyingStream
    */
    LineReadableStream.prototype.on = function (type, listener) {
        if (LineReadableStream._implementedEvents.indexOf(type) < 0)
            this._underlyingStream.on(type, listener);
        return _super.prototype.on.call(this, type, listener);
    };
    LineReadableStream.prototype.addListener = function (type, listener) {
        return this.on(type, listener);
    };

    LineReadableStream.prototype.removeListener = function (type, listener) {
        if (LineReadableStream._implementedEvents.indexOf(type) < 0)
            this._underlyingStream.removeListener(type, listener);
        return _super.prototype.removeListener.call(this, type, listener);
    };
    LineReadableStream.prototype.removeAllListeners = function (type) {
        if (LineReadableStream._implementedEvents.indexOf(type) < 0)
            this._underlyingStream.removeAllListeners(type);
        return _super.prototype.removeAllListeners.call(this, type);
    };

    LineReadableStream.prototype.resume = function () {
        if (this._underlyingStream.resume)
            this._underlyingStream.resume();
        return this;
    };

    LineReadableStream.prototype.pause = function () {
        if (this._underlyingStream.pause)
            this._underlyingStream.pause();
        return this;
    };

    /*
    // Does not seem to be available on Stream.Readable
    public destroy(): void
    {
    if (this._underlyingStream.destroy)
    this._underlyingStream.destroy();
    }
    */
    LineReadableStream.prototype.setEncoding = function (encoding) {
        if (this._underlyingStream.setEncoding)
            this._underlyingStream.setEncoding(encoding);
        return this;
    };
    LineReadableStream._implementedEvents = ["line", "end"];
    return LineReadableStream;
})(Stream.Readable);

module.exports = LineReadableStream;
